pkg/api/errors.go

```diff
@@ -1,44 +1,75 @@
 package api
 
-import "errors"
-
-var (
-	Error400     = errors.New("Bad Request")
-	Error401     = errors.New("Not Authorized")
-	Error403     = errors.New("Forbidden")
-	Error404     = errors.New("Resource Not Found")
-	Error409     = errors.New("Conflict")
-	Error412     = errors.New("Precondition Failed")
-	Error429     = errors.New("Too Many Requests")
-	Error500     = errors.New("API Internal Error")
-	ErrorUnknown = errors.New("Unknown API Error")
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 )
 
-func getStatusError(code int) error {
-	switch code {
-	case 200:
-		return nil
-	case 202:
+type APIError struct {
+	Error struct {
+		Status    int            `json:"status"`
+		Message   string         `json:"message"`
+		Errors    []ErrorMessage `json:"errors"`
+		Reference string         `json:"reference"`
+		Source    string         `json:"source"`
+	} `json:"error"`
+}
+
+type ErrorMessage struct {
+	Message string `json:"message"`
+}
+
+type BigAnimalError struct {
+	APIError APIError
+	Err      error
+}
+
+func (baerr *BigAnimalError) Error() string {
+	return fmt.Sprintf("status: %d - %v", baerr.APIError.Error.Status, baerr.APIError.Error.Message)
+}
+
+func (baerr *BigAnimalError) getDetails() string {
+	var details string
+	for _, err := range baerr.APIError.Error.Errors {
+		details += fmt.Sprintln(err.Message)
+	}
+	return details
+}
+
+func FromErr(err error) diag.Diagnostics {
+	if err == nil {
 		return nil
-	case 204:
+	}
+
+	baerr, ok := err.(*BigAnimalError)
+	if ok {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  fmt.Sprint(baerr.Err),
+				Detail:   baerr.getDetails(),
+			},
+		}
+	}
+	return diag.FromErr(err)
+}
+
+func getStatusError(code int, body []byte) error {
+	if code <= 299 {
 		return nil
-	case 400:
-		return Error400
-	case 401:
-		return Error401
-	case 403:
-		return Error403
-	case 404:
-		return Error404
-	case 409:
-		return Error409
-	case 412:
-		return Error412
-	case 429:
-		return Error429
-	case 500:
-		return Error500
-	default:
-		return Error500
+	}
+
+	apiErr := APIError{}
+
+	if json.Unmarshal(body, &apiErr) != nil {
+		return errors.New("unknown API Error")
+	}
+
+	return &BigAnimalError{
+		APIError: apiErr,
+		Err:      fmt.Errorf("status: %d - %v", apiErr.Error.Status, apiErr.Error.Message),
```

Comment:
There is a repetition with the Error string of the BigAnimalError above
```golang
func (baerr *BigAnimalError) Error() string {
	return fmt.Sprintf("status: %d - %v", baerr.APIError.Error.Status, baerr.APIError.Error.Message)
}
```
Is there a way to avoid this? 🤔

Reply:
I think we can just have 
```
type BigAnimalError struct {
	APIError APIError
}

func (baerr *BigAnimalError) Error() string {
	return fmt.Sprintf("status: %d - %v", baerr.APIError.Error.Status, baerr.APIError.Error.Message)
}

```

and then when we use it in the diags, we can just call it 
```
		return diag.Diagnostics{
			diag.Diagnostic{
				Severity: diag.Error,
				Summary:  baError.Error(),
				Detail:   baError.GetDetails(),
			},
		}
```

Reply:
Amazing! Thanks for the hint, @nstott! 🚀 🤘